firewood/src/merkle.rs

```diff
@@ -473,385 +440,356 @@ impl<T: WriteLinearStore> Merkle<T> {
-                Node::Leaf(last_node_leaf) => {
-                    // Replace the value in the leaf node.
-                    let new_leaf = Node::Leaf(LeafNode {
+                    self.update_node(
+                        traversal_path.iter().rev(),
+                        last_node.addr,
+                        updated_last_node,
+                    )?;
+
+                    return Ok(hash_invalidation_addresses);
+                };
+                // `last_node` is at a strict prefix of `key`.
+                // `remaining_key` is `key` after removing the prefix of `last_node` and 1 additional
+                // nibble for the child index that points to the node after `last_node`.
+                let remaining_key = &remaining_key[1..];
+
+                // See if the `last_node` has a child where `key` would go.
+                let child_option = *last_node_branch
+                    .children
+                    .get(last_node_to_child_index as usize)
+                    .expect("index is a nibble");
+
+                let Some(child_addr) = child_option else {
+                    // There is no child at the index that the `key` would be at.
+                    // Create a new leaf at that index.
+                    //     ...                ...
+                    //      |      -->         |
+                    //  last_node       updated_last_node
+                    //    /   \           /    |        \
+                    //  ...   ...       ...  new_leaf   ...
+                    let new_leaf_addr = self.create_node(Node::Leaf(LeafNode {
                         value,
-                        partial_path: last_node_leaf.partial_path.clone(),
-                    });
-
-                    let last_node_addr = last_node.addr;
-
-                    drop(last_node);
-                    match self.update_node(last_node_addr, new_leaf) {
-                        Ok(_) => {}
-                        Err(UpdateError::Io(e)) => return Err(e.into()),
-                        Err(storage::UpdateError::NodeMoved(new_addr)) => {
-                            // The leaf we're updating moved to a new address.
-                            // Update the leaf's parent to point to the new address.
-                            let last_node_parent = traversal_path.pop();
-
-                            let Some(leaf_parent) = last_node_parent else {
-                                // There is no parent of this leaf. It must be the root.
-                                self.set_root(new_addr)?;
-                                return Ok(hash_invalidation_addresses);
-                            };
+                        partial_path: Path::from_nibbles_iterator(remaining_key.iter().copied()),
+                    }))?;
+
+                    // Add new leaf as child of `last_node`.
+                    let mut updated_last_node_children = last_node_branch.children;
+                    *updated_last_node_children
```

Comment:
`use last_node_branch.mut_child`

Reply:
Not sure I follow. We're immutably borrowing `last_node_branch` so I don't think we can modify a child with `child_mut` here